src/main/java/org/sagebionetworks/bridge/services/PermissionService.java

```diff
@@ -0,0 +1,169 @@
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+import static org.sagebionetworks.bridge.models.permissions.EntityType.*;
+import static org.sagebionetworks.bridge.validators.PermissionValidator.INSTANCE;
+
+import com.google.common.collect.ImmutableSet;
+import org.sagebionetworks.bridge.BridgeUtils;
+import org.sagebionetworks.bridge.dao.PermissionDao;
+import org.sagebionetworks.bridge.exceptions.EntityNotFoundException;
+import org.sagebionetworks.bridge.models.accounts.Account;
+import org.sagebionetworks.bridge.models.accounts.AccountId;
+import org.sagebionetworks.bridge.models.accounts.AccountRef;
+import org.sagebionetworks.bridge.models.assessments.Assessment;
+import org.sagebionetworks.bridge.models.organizations.Organization;
+import org.sagebionetworks.bridge.models.permissions.EntityRef;
+import org.sagebionetworks.bridge.models.permissions.EntityType;
+import org.sagebionetworks.bridge.models.permissions.Permission;
+import org.sagebionetworks.bridge.models.permissions.PermissionDetail;
+import org.sagebionetworks.bridge.models.studies.Study;
+import org.sagebionetworks.bridge.validators.Validate;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+@Component
+public class PermissionService {
+    
+    private Set<EntityType> orgTypes = ImmutableSet.of(ASSESSMENT_LIBRARY, MEMBERS, ORGANIZATION, SPONSORED_STUDIES);
+    private Set<EntityType> studyTypes = ImmutableSet.of(PARTICIPANTS, STUDY, STUDY_PI);
+    private Set<EntityType> assessmentTypes = ImmutableSet.of(ASSESSMENT);
+    
+    private PermissionDao permissionDao;
+    
+    private AccountService accountService;
+    
+    private OrganizationService organizationService;
+    
+    private StudyService studyService;
+    
+    private AssessmentService assessmentService;
+    
+    @Autowired
+    final void setPermissionDao(PermissionDao permissionDao) {
+        this.permissionDao = permissionDao;
+    }
+    
+    @Autowired
+    final void setAccountService(AccountService accountService) {
+        this.accountService = accountService;
+    }
+    
+    @Autowired
+    final void setOrganizationService(OrganizationService organizationService) {
+        this.organizationService = organizationService;
+    }
+    
+    @Autowired
+    final void setStudyService(StudyService studyService) {
+        this.studyService = studyService;
+    }
+    
+    @Autowired
+    final void setAssessmentService(AssessmentService assessmentService) {
+        this.assessmentService = assessmentService;
+    }
+    
+    // For mock testing
+    protected String generateGuid() {
+        return BridgeUtils.generateGuid();
+    }
+    
+    public PermissionDetail createPermission(String appId, Permission permission) {
+        checkArgument(isNotBlank(appId));
```

Comment:
A null check is probably sufficient here. We're just looking for an internal error by a programmer

Reply:
Updated to just null checks.